### PR TITLE
2 small extra modules that helps to handle data (like JPA)

### PR DESCRIPTION
**Pull request by hendrikebbers**
Bean - Entity mapping
general service classes
basic JPA implementations

---
This change is [Reviewable](https://reviewable.io/reviews/canoo/dolphin-platform/835)

### CODE REVIEW

File: platform-extras/dolphin-platform-data-audit/src/main/java/com/canoo/dp/impl/platform/data/audit/AuditContext.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+public class AuditContext {
+
+    private final Javers javers;
+
+    public AuditContext(final ConnectionProvider connectionProvider, final String dialect, final String schema) {
+        this(connectionProvider, convertToDialect(dialect), schema);
+    }
+
+    public AuditContext(final ConnectionProvider connectionProvider, final DialectName dialect, final String schema) {
+        final JaversSqlRepository repository = SqlRepositoryBuilder.sqlRepository()
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-data-audit/src/main/java/com/canoo/dp/impl/platform/data/audit/AuditContext.java]: Maybe we should add some not-null checks here?

File: platform-extras/dolphin-platform-data-audit/src/main/java/com/canoo/dp/impl/platform/data/audit/AuditContext.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            public void onEntityUpdated(final AbstractEntity entity) {
+                if(isAuditable(entity)) {
+                    javers.commit(getAuthor(), entity);
+                }
+            }
+        });
+    }
+
+    private static DialectName convertToDialect(final String dialectName) {
+        if(dialectName.equals("h2")) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-data-audit/src/main/java/com/canoo/dp/impl/platform/data/audit/AuditContext.java]: Add a not-null check for dialectName or change the `if` statements to something like `"h2".equals(dialectName)`.
I would prefer to use some constants here too.

File: platform-extras/dolphin-platform-data-audit/src/main/java/com/canoo/dp/impl/platform/data/audit/AuditContext.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            return DialectName.POSTGRES;
+        }
+        if(dialectName.equals("mysql")) {
+            return DialectName.MYSQL;
+        }
+        throw new DolphinRuntimeException("Dialect not supported");
+    }
+
+    private <T extends  AbstractEntity> boolean isAuditable(final T entity) {
+        if(entity.getClass().isAnnotationPresent(Auditable.class)) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-data-audit/src/main/java/com/canoo/dp/impl/platform/data/audit/AuditContext.java]: Add not-null check here.

File: platform-extras/dolphin-platform-data-audit/src/main/java/com/canoo/dp/impl/platform/data/audit/AuditContext.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            return true;
+        }
+        return false;
+    }
+
+    private String getAuthor() {
+        return "unknown";
+    }
+
+    public <T extends  AbstractEntity> List<CdoSnapshot> getChanges(final T entity) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-data-audit/src/main/java/com/canoo/dp/impl/platform/data/audit/AuditContext.java]: Add not-null check here.

File: platform-extras/dolphin-platform-data-jpa/src/main/java/com/canoo/dp/impl/platform/data/jpa/AbstractEntity.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+import javax.persistence.MappedSuperclass;
+import javax.persistence.PostPersist;
+import javax.persistence.PostRemove;
+import javax.persistence.PostUpdate;
+import javax.persistence.Version;
+
+@MappedSuperclass
+public class AbstractEntity implements EntityWithId<Long> {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-data-jpa/src/main/java/com/canoo/dp/impl/platform/data/jpa/AbstractEntity.java]: Add `final` keyword to the method definitions.

File: platform-extras/dolphin-platform-data-jpa/src/main/java/com/canoo/dp/impl/platform/data/jpa/SimpleCrudService.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    @Override
+    public List<E> findAll() {
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+        CriteriaQuery<E> cq = cb.createQuery(entityType);
+        return entityManager.createQuery(cq).getResultList();
+    }
+
+    @Override
+    public E findById(Long id) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-data-jpa/src/main/java/com/canoo/dp/impl/platform/data/jpa/SimpleCrudService.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-data-jpa/src/main/java/com/canoo/dp/impl/platform/data/jpa/SimpleCrudService.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    @Override
+    public Class<E> getDataClass() {
+        return entityType;
+    }
+
+    @Override
+    public E save(final E entity) {
+        if (entity.getId() == null) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-data-jpa/src/main/java/com/canoo/dp/impl/platform/data/jpa/SimpleCrudService.java]: Maybe some not-null check should be added here.

File: platform-extras/dolphin-platform-data-jpa/src/main/java/com/canoo/dp/impl/platform/data/jpa/event/PersistenceListenerManager.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+import com.canoo.dp.impl.platform.data.jpa.AbstractEntity;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+/**
+ * Listener for all JPA events. Defined as a singleton. {@link PersistenceListener} instances can be registered as listener.
+ */
+public class PersistenceListenerManager {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-data-jpa/src/main/java/com/canoo/dp/impl/platform/data/jpa/event/PersistenceListenerManager.java]: Add `final` keyword to every method definition.

File: platform-extras/dolphin-platform-data/src/main/java/com/canoo/dp/impl/platform/data/i18n/I18nServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+    @Override
+    public String getText(String textKey, Locale locale) {
+        return getText(bundleBaseName, textKey, locale);
+    }
+
+    @Override
+    public String getText(String resourceBaseName, String textKey, Locale locale) {
+        ResourceBundle textBundle = Utf8ResourceBundle.getBundle(resourceBaseName, locale);
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-data/src/main/java/com/canoo/dp/impl/platform/data/i18n/I18nServiceImpl.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-data/src/main/java/com/canoo/dp/impl/platform/data/i18n/I18nServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+public class I18nServiceImpl implements I18nService {
+
+    private final String bundleBaseName;
+
+    public I18nServiceImpl(final String bundleBaseName) {
+        this.bundleBaseName = bundleBaseName;
+    }
+
+    @Override
+    public String getText(String textKey, Locale locale) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-data/src/main/java/com/canoo/dp/impl/platform/data/i18n/I18nServiceImpl.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-data/src/main/java/com/canoo/dp/impl/platform/data/i18n/I18nServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+        return getText(bundleBaseName, textKey, locale);
+    }
+
+    @Override
+    public String getText(String resourceBaseName, String textKey, Locale locale) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-data/src/main/java/com/canoo/dp/impl/platform/data/i18n/I18nServiceImpl.java]: Add `final` keyword.

**Comment by kunsingh** [comment]
Fixed review comments.

**Pull request merged**